shell: Support stdout/stderr redirect to a file

chu11 opened the pull request:
This PR adds support to the shell to output stdout/stderr to a file instead of writing to the KVS. All task stdout/stderr is written to a single file.

Notes:

- To set this, the user must set the `attributes.system.shell.options.output.<stream>.type` field to "file". The setting defaults to "kvs".

- The file to be written uses `attributes.system.shell.options.output.<stream>.path`.

- `attributes.system.shell.options.output.<stream>.label` will label each line of output with the task number it came from.

- I wasn't sure what the mode should be on files created. I just did 644 for now.

- Writes to files just calls `write(2)` everytime there is data to write. It'll block if the filesystem is busy. Could attach to a FD watcher instead?

chu11 commented on src/shell/task.c:
```diff
@@ -85,9 +86,9 @@ struct shell_task *shell_task_new (void)
 
-struct shell_task *shell_task_create (struct shell_info *info,
-                                      int index)
+struct shell_task *shell_task_create (flux_shell_t *shell, int index)
```
ahh, it's no longer necessary b/c "per-task"  output is all in `output.c` now

chu11 commented on src/shell/output.c:
```diff
@@ -464,11 +801,17 @@ static int shell_output_task_init (flux_plugin_t *p,
-                                              task_output_cb, out) < 0)
-        return -1;
+    if (output_type_requires_service (out->stdout_type)) {
+        if (flux_shell_task_channel_subscribe (task, "stdout",
```
I suppose so, although none of the currently supported ones will have that pattern.  Perhaps punt to when the situation arises?

chu11 commented:
during meeting discussion determined highest priority for this PR is

- stdout & stderr redirect going to single file
- adding special case mustache template of `{{id}}`

- If it can be completed in time, the nice to haves would be:
  - "amend RFC 24 to include a new redirect event for output streams" per @grondo comments above
  - non-blocking writes
  - but these can go in after this PR if necessary, need to create issues for them if they don't

chu11 commented:
re-pushed

- if stdout output type is specified, stderr defaults to it as well, including redirecting to the same file
- added mustanche template support for `{{id}}`
- fixed comment nit per @grondo comments
- fixed tests using `HAVE_JQ` pre-req

will try and do non-blocking writes tomorrow

garlick commented on src/cmd/flux-job.c:
```diff
@@ -922,7 +952,17 @@ void attach_exec_event_continuation (flux_future_t *f, void *arg)
+    }
+    else if (!strcmp (name, "output-kvs-ready")) {
+        /* Note that this event will only be output if atleast one of
+         * stdout/stderr will output to the kvs */
```
I opened #2399 for this idea.

chu11 commented:
re-pushed with some minor refactoring and fixing a mem-leak, just in case I can't finish non-blocking writes in a timely manner.

chu11 commented:
re-pushed with non-blocking write support.  Added support by using a `fd watcher` and storing off data to be written onto a list.  I dislike the fact I'm mallocing a new buffer to store this data, but seems like the only way to handle partial writes that can block.

Also did some cleanup along the way.  Hopefully all the code makes sense :-)

chu11 commented:
re-pushed, forgot a PATH_MAX check.

chu11 commented:
re-pushed, missed a chain-lint &&

chu11 commented:
re-pushed, minor cleanup in tests, I don't need to generate a new R1 & R2, I can just use the ones used by other tets.

chu11 commented:
restart builders, hit two spurious valgrinds

```
==10086== 
==10086== HEAP SUMMARY:
==10086==     in use at exit: 22,311 bytes in 59 blocks
==10086==   total heap usage: 101,119 allocs, 101,060 frees, 58,728,217 bytes allocated
==10086== 
==10086== LEAK SUMMARY:
==10086==    definitely lost: 0 bytes in 0 blocks
==10086==    indirectly lost: 0 bytes in 0 blocks
==10086==      possibly lost: 0 bytes in 0 blocks
==10086==    still reachable: 22,311 bytes in 59 blocks
==10086==         suppressed: 0 bytes in 0 blocks
==10086== Reachable blocks (those to which a pointer was found) are not shown.
==10086== To see them, rerun with: --leak-check=full --show-leak-kinds=all
```

chu11 commented:
oops, fixed a chain lint

grondo commented:
> added support by using a fd watcher and storing off data to be written onto a list. I dislike the fact I'm mallocing a new buffer to store this data, but seems like the only way to handle partial writes that can block.

One question -- would it be simpler to use a `flux_buffer_write_watcher` instead? (This is exactly its intended use anyway.)

chu11 commented:
Per discussion, re-did PR to output "redirect" events to the guest.output:

- Notes

- I output "0" if there's only 1 task, but "[0-N]" for > 1 task.
  Should I do brackets regardless?

- When a file is redirected, you can get a message like "0: stdout
  redirected to foo.out" by default.  A new '--quiet' option can
  suppress this output.

- There's a lot of tests, I'm going to put them in some new tests
  files.  If I can't finish that by the end of work, I'll do a push
  when I get home.

chu11 commented:
re-pushed, all the new tests are in two new files now.  Also, one additional note I forgot to mention.  If both stdout & stderr are redirected to the same file, `flux job attach` outputs two messages indicating the redirection:

```
[0-1]: stdout redirected to foo
[0-1]: stderr redirected to foo
```

I couldn't think of an obvious clean way to deal with for the time being.

chu11 commented:
re-pushed, fixed a mem-leak code coverage found

garlick commented on src/cmd/flux-job.c:
```diff
@@ -829,6 +844,39 @@ void attach_output_continuation (flux_future_t *f, void *arg)
+        if (!strcmp (stream, "stdout"))
+            ctx->stdout_redirect_count += idset_count (idset);
+        else
+            ctx->stderr_redirect_count += idset_count (idset);
```
It seems like this code shouldn't actually need to count redirects and cancel the RPC.  The RPC stream will eventually end with ENODATA when the job ends.

Since we're ignoring EOFs it seems like we can ignore these too, except for printing the warning message.

The block comment at the top of the function should be updated to mention redirect events.

chu11 replied:
ahh yes, i had done things a little differently in a earlier incarnation of the code.

garlick commented on src/shell/output.c:
```diff
@@ -63,8 +63,16 @@ enum {
+    int len;
+    int offset;
+};
+
```
Is the "buffer watcher" not useful here?  E.g. `flux_buffer_write_watcher_create()`?

chu11 replied:
main issue was buffer watcher has to have a buffer size set when it's created.  So I elected to go with the fd watcher.  

Thinking this through, could possible save data from `out->output` somewhere else when buffer is full.  Not sure if it's a net win code wise.

It did just occur me that I should probably add a completion reference for fd-watcher writes, since on exit, there's no way to know if there's still data for the fd watcher to keep on writing.    As a side effect, would clean up some of the `shell_output_destroy()` path.

garlick replied:
Ok, I'll have a closer look at this code then.  Are there tests that can make write(2) return EAGAIN/EWOULDBLOCK (seems like that would be hard to do if output is always a file).

chu11 replied:
Nah, no tests for that specific path.  Hard to create a test.

BUT, I only checked for EAGAIN, totally forgot about EWOULDBLOCK.  Need to fix that up.

chu11 replied:
oh wait, EWOULDBLOCK is only for sockets.  I had it right :-)

garlick replied:
Wait a second, the data is already "buffered" in the `output` JSON array, so we shouldn't need to buffer it again in the zlist.  It should be possible to get this done by just adding a fd watcher and an offset count for each of stderr/stdout:
* init offset to zero
* start fd watcher (POLLIN) on transition of output array from empty to non-empty
* on fd callback, decode "top" element of `output` and write(2) data+offset, len-offset
* if EAGAIN/EWOULDBLOCK, return (no-op)
* if full write, pop element, zero offset, and return (stop watcher if output is now empty)
* if partial write, add size to offset and keep writing until EAGAIN/EWOULDBLOCK

It may seem ugly for the case where you get EAGAIN/EWOULDBLOCK after a partial write to decode the same element multiple times.  However, it seems like that could be a future optimization if that code path is found to be "hot".  Its hard to know if it will be.

Also, the code to write to stdout/stderr for the standalone case could use this code path with fds set to STDOUT_FILENIO/STDERR_FILENO.

grondo replied:
> seems like that would be hard to do if output is always a file 

Yeah, I wonder how hard it would be to allow output to stdin (using `popen`) of a process instead of a file (for testing only). It would be nice, but probably a fair bit of code just for testing.

>  main issue was buffer watcher has to have a buffer size set when it's created. So I elected to go with the fd watcher.

I haven't looked at the code, but I guess it makes sense when you already have the  data buffered elsewhere (in the output json array here)

However, the underlying buffer *is* a cbuf, which can grow if necessary (if we exposed those hooks).

garlick replied:
> Also, the code to write to stdout/stderr for the standalone case could use this code path with fds set to STDOUT_FILENIO/STDERR_FILENO.

Which could be wired to a pipe or something to simulate EAGAIN/partial write handling as was done in librouter/test/sendfd.c.

grondo replied:
Great suggestion!

garlick replied:
Just saw @grondo's suggestion to use the write watcher.  Missed that earlier, so sorry for the redundant comment.

grondo replied:
No problem!

I really like generalizing the "file" output to "fd" output and switch terminal output to that same handler. However, I wonder if we are running out of time for that approach in this PR?

We could delay merging the PR and pursue this cleanup, or merge the (working) PR soon and do the cleanup as a follow-on PR. (I also kind of want to investigate moving the "output modes" into their own plugins, but haven't thought to hard about how that would work yet. But the output plugin could change regardless..)

I leave the approach up to @chu11 and @garlick

garlick replied:
That works for me.  @chu11, if that makes sense to you, then my vote would be to drop the "support nonblocking writes" commit, and then this is probably good to go in IMHO.

chu11 replied:
> Wait a second, the data is already "buffered" in the output JSON array, so we shouldn't need to buffer it again in the zlist. 

So the main reason I copy to another list is b/c both stdout and stderr both go to the same output JSON array, but the streams could go to different locations, like file/kvs/terminal. or different files  One stream could be blocking and one not.

If I leave everything on the output array, then I have to iterate through the entire array everytime, even if one stream isn't blocked.  Perhaps this isn't the end of the world and not as big a deal as I originally thought.  Some logic has to be changed since I can't call `json_array_clear()` anymore, gotta remove elements from the output array as they are processed.

> if partial write, add size to offset and keep writing until EAGAIN/EWOULDBLOCK

(Note, I now see a mistake I did in `output.c`, so following sentence is in principle) One of the side reasons I did the list is b/c when you call `iodecode()` the `data` from the log entry is malloced when returned to the caller.  This decode+malloc has to be done.  It seemed like little cost to save off this already-malloced region of memory, rather than leave it on the `output` array to be re-malloced the next time your try to read from the array.

garlick replied:
Oh I missed that detail, sorry.  The streams mixed up in one array does make my proposal a bit trickier.

chu11 replied:
> That works for me. @chu11, if that makes sense to you, then my vote would be to drop the "support nonblocking writes" commit, and then this is probably good to go in IMHO.

Ok, let me remove that commit and I'll re-push

garlick commented on t/t2606-job-shell-output-redirection.t:
```diff
@@ -0,0 +1,447 @@
+        id=$(cat j1echostdout-0 | flux job submit) &&
+        flux job wait-event $id clean &&
+        grep stdout:foo out0
+'
```
Now that `flux mini run` has been merged, and has options for file redirection, I think many of these tests can be shortened and HAVE_JQ dropped, e.g. the above could be
```shell
flux mini run -n1 --output=out0 ${TEST_SUBPROCESS_DIR}/test_echo -P -O foo &&
grep stdout:foo out0
```
In addition to `--output` and `--error`, there is also `-o, --setopt` to directly set shell option attributes.

chu11 commented:
re-based and re-pushed.

- removed redirect counts in `flux-job` per comments above and adjusted some comments
- added a completion ref and some comments for the fd watcher in shell `output.c`
- updated tests in `t2606` to use `flux-mini`.

chu11 commented:
re-pushed, removing non-blocking writes per discussion above.

Shall we create an issue to track the blocking issue?

garlick commented:
Yep good idea!

On Sat, Sep 28, 2019, 1:28 PM Al Chu <notifications@github.com> wrote:

> re-pushed, removing non-blocking writes per discussion above.
>
> Shall we create an issue to track the blocking issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2395?email_source=notifications&email_token=AABJPWZB3AK6EGWEE72LDJLQL645VA5CNFSM4I2QGC5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD73BZRY#issuecomment-536222919>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPW2UJUG64EFOFWDH3HTQL645VANCNFSM4I2QGC5A>
> .
>
